### PR TITLE
Refactor Initialized Model Export

### PR DESCRIPTION
**Pull request by Satrat**
* Fixing bug where exporting a model initialized directly from the HuggingFace hub failed
* Moved config updates for kv-cache injection to the export function, and add support for kv-cache injection export to instantiated models
* Automatically infer transformers integration based on model properties

### Example
```python
from sparseml.transformers import SparseAutoModelForCausalLM, SparseAutoTokenizer
from sparseml.transformers import oneshot
from sparseml.modifiers import SparseGPTModifier
from sparseml import export

ORIG_MODEL_PATH = "roneneldan/TinyStories-1M"
OUTPUT_DIR = "./test_output_here"

model = SparseAutoModelForCausalLM.from_pretrained(ORIG_MODEL_PATH, device_map="auto")
tokenizer = SparseAutoTokenizer.from_pretrained(ORIG_MODEL_PATH)

oneshot(
    model=model,
    tokenizer=tokenizer,
    dataset="open_platypus",
    recipe=SparseGPTModifier(sparsity=0.5),
    output_dir="./test_export_output",
    num_calibration_samples=512,
)

export(
    model=model,
    tokenizer=tokenizer,
    target_path=OUTPUT_DIR
)
```

### CODE REVIEW

File: tests/sparseml/export/transformers/test_generative_transformers.py [hunk 1]
```diff
@@ -49,21 +46,11 @@ def test_export_initialized_model_no_source_path(self, setup):
...
+            model=SparseAutoModelForCausalLM.from_pretrained(source_path),
+            tokenizer=SparseAutoTokenizer.from_pretrained(source_path),
             target_path=target_path,
-            integration="transformers",
             sequence_length=384,
-            # we need to disable applying kv cache injection
-            # because the script does not have access to the
-            # config.json (we are not creating a full deployment
-            # directory during the export)
-            graph_optimizations="none",
```
Comment (dbogunowicz) [inline tests/sparseml/export/transformers/test_generative_transformers.py:66]: excellent, much cleaner and better UI

File: src/sparseml/pytorch/torch_to_onnx_exporter.py [hunk 2]
```diff
@@ -129,7 +129,8 @@ def remove_leftover_files(self):
             torch_onnx_export_transform, _TorchOnnxExport
         ), "Expected the first transform from self.transform to be _TorchOnnxExport"
         for file in torch_onnx_export_transform.leftover_files:
-            os.remove(file)
+            if os.path.exists(file):
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/torch_to_onnx_exporter.py:132]: just curious, why is this change needed now?

File: src/sparseml/integration_helper_functions.py [hunk 3]
```diff
@@ -37,23 +37,29 @@ class Integrations(Enum):
...
+    :param source_model: An instantiated model to export
     :param integration: Optional name of the integration to use. If not provided,
         will attempt to infer it from the source_path.
     :return: The name of the integration to use for exporting the model.
     """
 
     integration = integration or _infer_integration_from_source_path(source_path)
 
+    # attempt to infer transformers based on model attribute
+    if source_model is not None and hasattr(source_model, "config_class"):
```
Comment (dbogunowicz) [inline src/sparseml/integration_helper_functions.py:60]: why `config_class` is a deciding attribute here?

**Review by dbogunowicz — APPROVED** [review]
Left a few comments, all in all, great changes!

**Review by bfineran — APPROVED** [review]

File: src/sparseml/integration_helper_functions.py (hunk 3 above)
Comment (Satrat) [reply src/sparseml/integration_helper_functions.py:60]: This is really just to tell if its a PreTrainedModel, but I didn't want to have to add the transformers dependency to this part of the repo

File: src/sparseml/export/export.py [hunk 4]
```diff
@@ -206,8 +206,23 @@ def export(
...
             target_path = source_path
+        if tokenizer is not None:
+            _LOGGER.warning(
+                "Passed a tokenizer is not supported when exporting from ",
+                "a source path. The tokenizer will be ignored. ",
+            )
 
-    integration = resolve_integration(source_path, integration)
+    if model is not None:
+        model.config = remove_past_key_value_support_from_config(model.config)
```
[the remark this answers is not in the document]
Comment (Satrat) [reply src/sparseml/export/export.py:216]: Good point, I was thinking too transformers focused :) will update
